File: eratosthenes/eratosthenes3.cc
#include "eratosthenes3.h"

#include <cmath>

namespace {
#include "constants.h"
}  // namespace

void Eratosthenes3::generate(int64 x) {
  if (x > 10000000000)
    return;

  generateSmall(std::ceil(std::sqrt(x)));
  int64 size = x / 30 + (x % 30 != 0);
  initFlags(x, size);
  for (uint8* segment = flags_.data(); size > 0;
       segment += kSegmentSize, size -= kSegmentSize) {
    generateCore(segment, std::min<uint64>(size, kSegmentSize));
  }
}

void Eratosthenes3::initFlags(const int64 x, const int64 size) {
  flags_.clear();
  flags_.resize(size, 0xff);

  flags_[0] = 0xfe;
  if (int r = x % 30) {
    if (r <= 1)       flags_[size - 1] = 0x0;
    else if (r <= 7)  flags_[size - 1] = 0x1;
    else if (r <= 11) flags_[size - 1] = 0x3;
    else if (r <= 13) flags_[size - 1] = 0x7;
    else if (r <= 17) flags_[size - 1] = 0xf;
    else if (r <= 19) flags_[size - 1] = 0x1f;
    else if (r <= 23) flags_[size - 1] = 0x3f;
    else if (r <= 29) flags_[size - 1] = 0x7f;
  }
}

void Eratosthenes3::generateSmall(const int64 sqrt_x) {
  sflags_.clear();
  indecies_.clear();

  const uint64 sqrt_xi = sqrt_x / 30 + 1;
  const int64 quart_x = std::ceil(std::sqrt(sqrt_x));
  const uint64 quart_xi = quart_x / 30 + 1;

  sflags_.resize(sqrt_xi, 0xff);
  sflags_[0] = 0xfe;
  for (uint64 i = 0; i < quart_xi; ++i) {
    for (uint8 flags = sflags_[i]; flags; flags &= flags - 1) {
      uint8 lsb = flags & (-flags);
      int ibit = BitToIndex(lsb);
      const int32 m = kMod30[ibit];
      int32 pm = 30 * i + 2 * m;
      uint64 j = i * pm + (m * m) / 30;
      uint64 k = ibit;
      while (j < sflags_.size()) {
        sflags_[j] &= kMask[ibit][k];
        j += i * C1[k] + C0[ibit][k];
        k = (k + 1) & 7;
      }
    }
  }

  for (uint64 i = 0; i < sflags_.size(); ++i) {
    for (uint8 flags = sflags_[i]; flags; flags &= flags - 1) {
      uint8 lsb = flags & (-flags);
      int ibit = BitToIndex(lsb);
      const int32 m = kMod30[ibit];
      uint64 j = i * (30 * i + 2 * m) + (m * m) / 30;
      indecies_.push_back(((j + kSegmentSize) << 3) | ibit);
    }
  }
}

void Eratosthenes3::generateCore(uint8* flags, const uint64 size) {
  auto p_index = indecies_.begin();
  for (uint64 i = 0; i < sflags_.size(); ++i) {
    for (uint8 primes = sflags_[i]; primes; primes &= primes - 1) {
      uint8 lsb = primes & (-primes);
      int ibit = BitToIndex(lsb);
      const int32 m = kMod30[ibit];
      uint64 index = *p_index;
      uint64 j = (index >> 3) - kSegmentSize;
      uint64 k = index & 7;
      while (j < size) {
        flags[j] &= kMask[ibit][k];
        j += i * C1[k] + C0[ibit][k];
        k = (k + 1) & 7;
      }
      *p_index = (j << 3) | k;
      ++p_index;
    }
  }
}

int64 Eratosthenes3::count() {
  if (flags_.empty())
    return -1;

  int64 ret = 3;  // count 2, 3, 5
  for (uint8 f : flags_) {
    ret += PopCnt(f);
  }
  return ret;
}

File: eratosthenes/rust/src/eratosthenes/eratosthenes3.rs
use crate::bits::Bytes;

use super::{Eratosthenes, BIT_MASKS, C0, C1, MOD30, SEGMENT_SIZE};

pub struct Eratosthenes3 {
    flags: Bytes,
    sflags: Bytes,
    indecies: Vec<u32>,
}

impl Eratosthenes3 {
    pub fn new() -> Eratosthenes3 {
        Eratosthenes3 {
            flags: Bytes::new(0, 0),
            sflags: Bytes::new(0, 0),
            indecies: Vec::new(),
        }
    }

    fn generate_small(&mut self, sx: usize) {
        let sxi = sx / 30 + 1;
        let qx = (sx as f64).sqrt().ceil() as usize;
        let qxi = qx / 30 + 1;

        self.sflags = Bytes::new(sxi, 0xff);
        let sflags = &mut self.sflags;
        sflags.unset_bit(0, 0);
        for i in 0..qxi {
            let mut bits = sflags.get_byte(i);
            while bits != 0 {
                let lsb = bits & (!bits + 1);
                let ibit = (lsb - 1).count_ones() as usize;
                let m = MOD30[ibit];
                let pm = 30 * i + 2 * m;

                let mut j = i * pm + (m * m) / 30;
                let mut k = ibit;
                while j < sxi {
                    sflags.mask(j, BIT_MASKS[ibit][k]);
                    j += i * C1[k] + C0[ibit][k];
                    k = (k + 1) & 7;
                }
                bits &= bits - 1;
            }
        }

        let indecies = &mut self.indecies;
        indecies.clear();
        for i in 0..sxi {
            let mut bits = sflags.get_byte(i);
            while bits != 0 {
                let lsb = bits & (!bits + 1);
                let ibit = (lsb - 1).count_ones() as usize;
                let m = MOD30[ibit];
                let j = i * (30 * i + 2 * m) + (m * m) / 30;
                indecies.push(((j << 3) | ibit) as u32);
                bits &= bits - 1;
            }
        }
    }

    fn init_flags(&mut self, x: usize, size: usize) {
        let mut flags = Bytes::new(size, 0xff);

        flags.set_byte(
            size - 1,
            match x % 30 {
                0 => 0xff,
                1 => 0x00,
                2..=7 => 0x01,
                8..=11 => 0x03,
                12..=13 => 0x07,
                14..=17 => 0x0f,
                18..=19 => 0x1f,
                20..=23 => 0x3f,
                _ => 0x7f,
            },
        );
        flags.unset_bit(0, 0); // 1 is not a prime

        self.flags = flags;
    }

    fn generate_core(&mut self, offset: usize, size: usize) {
        let flags = &mut self.flags;
        let mut pindex = 0;
        for i in 0..self.sflags.len() {
            let mut primes = self.sflags.get_byte(i);
            while primes != 0 {
                let lsb = primes & (!primes + 1);
                let ibit = (lsb - 1).count_ones() as usize;
                let index = self.indecies[pindex] as usize;
                let mut j = index >> 3;
                let mut k = index & 7;
                while j < size {
                    flags.mask(j + offset, BIT_MASKS[ibit][k]);
                    j += i * C1[k] + C0[ibit][k];
                    k = (k + 1) & 7;
                }
                j -= size;
                self.indecies[pindex] = ((j << 3) | k) as u32;
                pindex += 1;
                primes &= primes - 1;
            }
        }
    }
}

impl Eratosthenes for Eratosthenes3 {
    fn version(&self) -> i32 {
        3
    }

    fn generate(&mut self, x: usize) {
        let sqr_x = (x as f64).sqrt().ceil() as usize;
        self.generate_small(sqr_x);

        let r = x % 30;
        let size = x / 30 + if r == 0 { 0 } else { 1 };
        self.init_flags(x, size);

        for offset in (0..size).step_by(SEGMENT_SIZE) {
            let segment_size = (size - offset).min(SEGMENT_SIZE);
            self.generate_core(offset, segment_size);
        }
    }

    fn generate_range(&mut self, _x: usize, _y: usize) {
        eprintln!("Does not support ranged generation.");
    }

    fn count(&self) -> usize {
        // 2, 3, and 5 are counted separately.
        3 + self.flags.bit_count()
    }
}
